### ISSUE TITLE
Ideas for firewall circumvention

### ISSUE THREAD

**Issue by silversquirl** [softminus/hellcat#3]
Currently, two ports are needed for hellcat to operate: a control port and a data port. This can cause issues in firewalled environments where very few ports are accessible. For example, I run an SSH daemon on port 443 so that I can access it from behind a firewall that blocks all outgoing connections except for HTTP and HTTPS. This means that I only have one port left (80) and so cannot use hellcat effectively.

I have two ideas of how to get around this, but there are probably more ways it could be done:

  1. Allow the hosts for the control and data ports to be different
     This would allow SSH port forwarding to be used for the control port, making my usecase possible.
  2. Remove the need for a control port altogether.
     From a quick look at the source code, it seems that the control port simply tells the receiver when to stop receiving. Perhaps this could be done by closing a data port connection before `chunksize` bytes have been sent. That way, if the data being transferred is divisible by `chunksize` a new connection could be opened and then immediately closed when the transfer is finished, and if not, the connection would be automatically closed.

I've not had a detailed look at the code yet, so I don't know whether or not these ideas would work, but I'll try and take a look at some point, and I may try and patch this functionality in.

Let me know what you think and if you have any other suggestions!

**Comment by vktec** [comment]
I've made a patch for a separate `controlport` and will submit a PR now. Still looking into the possibility of removing the need for the controlport altogether.

**Comment by vktec** [comment]
Of course, I can't actually *push* my PR because I'm on a firewalled connection, lol.
I'll start looking at my second idea and push when I get home.

**Comment by vktec** [comment]
I've managed to remove the need for the `controlport`, so both my ideas do work.
From the README, it seems there is planned feature to allow `hellsend` to tell `hellrecv` what the `chunksize` is. Obviously some kind of transport would be needed to do this, which would be the `controlport`.
With the `controlport` removed, this could be acheived by sending a single integer of a set length (say, 16-bit or 32-bit unsigned) before sending any data, and using that as the `chunksize`. Of course, there may be issues with endianness, but these would occur with the `controlport` as well and are relatively easy to solve using a few bitwise operations.

**Comment by vktec** [comment]
Okay, I've got two branches ready to PR, so I'm wondering which you'd prefer: the one that adds `controlhost` or the one that removes `controlport`.
